File: DynamicProgramming/Leetcode/Medium/HouseRobber/HouseRobber.java
// https://leetcode.com/problems/house-robber/

import java.util.Arrays;

public class HouseRobber {
    // TABULATION
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        int prev1 = nums[0], prev2 = 0;
        for (int i = 1; i < nums.length; i++) {
            int take = nums[i] + prev2;
            int notTake = prev1;
            prev2 = prev1;
            prev1 = Math.max(take, notTake);
        }
        return prev1;
    }
    // MEMOIZATION
    int jump(int[] nums, int ind, int dp[]) {
        if (ind < 0) {
            return 0;
        }
        if (dp[ind] != -1) {
            return dp[ind];
        }
        // rob
        int left = jump(nums, ind - 2, dp) + nums[ind];
        // leave
        int right = jump(nums, ind - 1, dp);
        dp[ind] = Math.max(left, right);
        return dp[ind];
    }
    public int rob2(int[] nums) {
        if(nums.length == 0) {
            return 0;
        }
        if(nums.length == 1) {
            return nums[0];
        }
        int dp[] = new int[nums.length];
        Arrays.fill(dp, -1);
        return jump(nums, 0, dp);
    }
}

File: DynamicProgramming/Leetcode/Medium/HouseRobber/HouseRobber.cpp
// https://leetcode.com/problems/house-robber/
#include<bits/stdc++.h>
using namespace std;
// TABULATION
int rob(vector<int> &nums)
{
    if (nums.size() == 0)
    {
        return 0;
    }
    if (nums.size() == 1)
    {
        return nums[0];
    }
    int prev1 = nums[0], prev2 = 0;
    for (int i = 1; i < nums.size(); i++)
    {
        int take = nums[i] + prev2;
        int notTake = prev1;
        prev2 = prev1;
        prev1 = max(take, notTake);
    }
    return prev1;
}
// MEMOIZATION
int jump(vector<int> &nums, int ind, vector<int> &dp)
{
    if (ind < 0)
    {
        return 0;
    }
    if (dp[ind] != -1)
    {
        return dp[ind];
    }
    // rob
    int left = jump(nums, ind - 2, dp) + nums[ind];
    // leave
    int right = jump(nums, ind - 1, dp);
    dp[ind] = max(left, right);
    return dp[ind];
}
int rob2(vector<int> &nums)
{
    if (nums.size() == 0)
    {
        return 0;
    }
    if (nums.size() == 1)
    {
        return nums[0];
    }
    vector<int> dp(nums.size(), -1);
    return jump(nums, nums.size() - 1, dp);
}
int main(void) {
    
    return 0;
}